lib/chef/resource/locale.rb

```diff
@@ -25,67 +25,121 @@ class Locale < Chef::Resource
-            # RHEL uses /etc/locale.conf
-            not_if { up_to_date?("/etc/locale.conf", new_resource.lang) } if ::File.exist?("/etc/locale.conf")
-            # Ubuntu 16.04 still uses /etc/default/locale
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang) } if ::File.exist?("/etc/default/locale")
+        begin
+          unless up_to_date?
+            converge_by "Updating System Locale" do
+              generate_locales unless unavailable_locales.empty?
+              update_locale
+            end
           end
-        elsif ::File.exist?("/etc/sysconfig/i18n")
-          locale_file_path = "/etc/sysconfig/i18n"
+        rescue
+          # It might affect debugging
+          raise "#{node['platform']} platform is not supported by the chef locale resource. " +
+            "If you believe this is in error please file an issue at https://github.com/chef/chef/issues"
+        end
+      end
 
-          updated = up_to_date?(locale_file_path, new_resource.lang, new_resource.lc_all)
+      action_class do
 
-          file locale_file_path do
-            content lazy {
-              locale = IO.read(locale_file_path)
-              variables = Hash[locale.lines.map { |line| line.strip.split("=") }]
-              variables["LANG"] = new_resource.lang
-              variables["LC_ALL"] =
-                variables.map { |pairs| pairs.join("=") }.join("\n") + "\n"
-            }
-            not_if { updated }
+        # Generates the localisation files from templates using locale-gen.
+        # @see http://manpages.ubuntu.com/manpages/cosmic/man8/locale-gen.8.html
+        # @raise [Mixlib::ShellOut::ShellCommandFailed] not a supported language or locale
+        #
+        def generate_locales
+          bash "Generating locales: #{unavailable_locales.join(' ')}" do
+            code <<~CODE
+              if type locale-gen >/dev/null 2>&1
+              then
+                locale-gen #{unavailable_locales.join(' ')}
+              fi
+            CODE
           end
+        end
 
-          execute "reload root's lang profile script" do
-            command "source /etc/sysconfig/i18n; source /etc/profile.d/lang.sh"
-            not_if { updated }
+        # Updates system locale by appropriately writing locales in /etc/locale.conf
+        # @note Also sourcing to that file with export to update thier values on current shell
+        # @see https://wiki.archlinux.org/index.php/locale#Setting_the_system_locale
+        #
+        def update_locale
+          file "Updating system locale" do
+            path LOCALE_CONF
+            content new_content
           end
-        elsif ::File.exist?("/usr/sbin/update-locale")
-          execute "Generate locales" do
-            command "locale-gen"
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang, new_resource.lc_all) }
+          bash "Reload root's lang profile script" do
```

Comment:
This won't affect the current environment, so we should drop this bash resource.

As discussed, there are lots of edge cases here. Like the chef-client update cookbook, we could try to kill chef-client and hope that our init is systemd or something that will restart us, but at this point I'd leave that exercise to the user.

Add a comment to the `update_local` property like: "This locale change won't affect the current run. At this time it is an exercise left to the user to restart or reboot if the locale change is required a later part of the client run."